### ISSUE TITLE
View records with non-string id cannot be fetched

### ISSUE THREAD

**Issue by abrman** [pocketbase/pocketbase#3110]
While this is by no means a breaking bug and everything works as intended, perhaps hiding preview if broken would be one way to solve this. :)

![image](https://github.com/pocketbase/pocketbase/assets/6663260/d255bd83-a7cc-4166-8ca2-48b23aa7f36c)

**Comment by ganigeorgiev** [comment]
I'm not able to reproduce it locally or on the live demo with v0.17.3.

Make sure that the file actually exists (eg. if you migrated from local to S3 you have to manually upload to S3 your old files).

If that's not the case, please paste the error message from the browser inspect tool and I'll try to investigate it further.

**Comment by abrman** [comment]
v0.17.3, no S3 bucket (local sotrage), preview works in original collection `users`
Error: `GET http://127.0.0.1:8090/api/files/hc392vhwl9tvcqi/1/blob_QGte3QLCrp.png?thumb=100x100&token= 404 (Not Found)`

![image](https://github.com/pocketbase/pocketbase/assets/6663260/52d7da73-b81a-40e5-8e59-2d10330b91e8)

**Comment by ganigeorgiev** [comment]
I'm not sure how to reproduce it. Is the file in the non-view collection preview-able?

**Comment by ganigeorgiev** [comment]
Could you also paste your View query?

**Comment by abrman** [comment]
![image](https://github.com/pocketbase/pocketbase/assets/6663260/97d0d24f-9087-4c59-b1fa-0778265bb80b)
```sql
SELECT
    (ROW_NUMBER() OVER()) as id,
    p.id AS project_id,
    u.id AS user_id,
    u.display_name AS display_name,
    u.avatar AS avatar
FROM
    Projects p
JOIN
    users u ON u.projects LIKE '%' || p.id || '%';
```

**Comment by abrman** [comment]
Some of myobservations:
- Working preview URL: `http://127.0.0.1:8090/api/files/users/3jopq76jnhxojno/blob_QGte3QLCrp.png`
- What's being shown: `http://127.0.0.1:8090/api/files/hc392vhwl9tvcqi/1/blob_QGte3QLCrp.png?token=`

The `1` in the URL is being taken from `(ROW_NUMBER() OVER()) as id` and I'm not sure if `users` collection's ID is `hc392vhwl9tvcqi`, or it's the id for the View Collections, but replacing the two with `users` and actual users ID gets the image to load.

**Comment by abrman** [comment]
![image](https://github.com/pocketbase/pocketbase/assets/6663260/46b8a36c-1486-46c9-a962-2184e2cd4f7d)

**Comment by abrman** [comment]
`--debug` seems to show the same thing.
Apart of `[0.00ms] SELECT .*` rows logs, just one stands out: `2023/08/10 20:30:19 sql: no rows in result set`

**Comment by abrman** [comment]
Have you been able to reproduce? If not I'll get a minimal working `pb_data` zip (one where bug is observable). If it makes any difference I'm running on windows.

**Comment by ganigeorgiev** [comment]
I was able to reproduce it. It seems that there is something with the view record retrieval that doesn't work for some reason. I'm currently investigating it.

**Comment by ganigeorgiev** [comment]
The integer id is causing it to break. You can try to change your view id to a string based one, eg.:

```sql
(ROW_NUMBER() OVER() || '') as id
-- or
CAST(ROW_NUMBER() OVER() as TEXT) as id
```

I'll see if it can be normalized on a query level or if it ended up inefficient to just update the ID note in the View UI.

**Comment by ganigeorgiev** [comment]
This is a little more complicated than I anticipated. I have a cold and cannot think properly at the moment and will leave it for tomorrow or for the weekend.

Views with int ids are kind of broken currently. Internally we don't support non-string collection primary keys and while **in the response the id is always returned as string** because that's the type of the model field, internally SQLite store it as int value which causes the direct id=:param equal comparison to fail.

In general I have in my todo to allow customizing the `id` field (and any other system non-schema field) for more "native" SQLite experience, but for now this remains a low priority as it requires a lot of internal changes (_I'll reconsider prioritizing it sometime after the logs generalization_).

In the meantime, we'll have to:
- check if possible to throw a validation error for non-int view ids (_this may not be possible to detect and validate because expressions are hard to determine upfront what type will yield and even calling `table_info` could return empty string/null for such columns; on top of that SQLite by default allows storing mixed value types in a single column_)
- update the view api tests (including the view/update/delete actions) to check the int id scenario
- create a system migration that will attempt to normalize any `(ROW_NUMBER() OVER()) id` view query statements as `(ROW_NUMBER() OVER() || '') id`

**Comment by ganigeorgiev** [comment]
I think the string normalization can be done behind the scenes without breaking or requiring user changes by creating a wrapping plain `SELECT` query over the view query, aka.:

```sql
SELECT cast(id as text) id, ... FROM (view_query)
```

When the `id` column refers to id from a regular collection,  the query planner indicates that the related covering `id` index is preserved and still being used even with the cast.
This is further confirmed from my local tests with the [benchmarks repo dataset](https://github.com/pocketbase/benchmarks), meaning that the above wrapping doesn't seem to have significant impact on the performance even on the larger posts100k collection.
This is probably because of the [builtin SQLite Subquery flattening optimization](subquery_flattening).

I'll spend a little more time testing it further sometime later today and will update the automated test cases.

**Comment by ganigeorgiev** [comment]
I've submitted a fix in `master` and it will be available with the minor next v0.17.4 release sometime later today.

The view record retrieval should work even without the manual text cast. We perform behind the scene query wrapping only for the cases where we can't identify the column type (expressions, bool, numeric values, etc.).
